src/server/pkg/storage/fileset/api.go

```diff
@@ -0,0 +1,29 @@
+package fileset
+
+import (
+	"io"
+
+	"github.com/pachyderm/pachyderm/src/server/pkg/storage/fileset/index"
+	"github.com/pachyderm/pachyderm/src/server/pkg/storage/fileset/tar"
+)
+
+// File represents a file in a fileset
+type File interface {
+	// Index returns the index for the file
+	Index() *index.Index
+	// Header returns a *tar.Header with metadata about the file, or an error
+	Header() (*tar.Header, error)
+	// Content writes the content of the file to w
+	Content(w io.Writer) error
+}
+
+var _ File = &FileMergeReader{}
+var _ File = &FileReader{}
+
+type FileSource interface {
```

Comment:
Maybe this should just be `FileIterator`, since that is what it does?

Reply:
`Iterator` and `Reader` both seem stateful to me.  Iterator in most other contexts, java, python etc. is a stateful object with a next or hasNext method. I think the eventual behavior here should be that Iterate can be called as many times as you want.  The MergeReader and Reader do not support that yet.